### DISCUSSION TITLE
Heatmaps - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#6873] [Docs]
# Heatmaps - Ultralytics YOLOv8 Docs

Advanced Data Visualization with Ultralytics YOLOv8 Heatmaps

https://docs.ultralytics.com/guides/heatmaps/

**Comment by RizwanMunawar** [comment]
Heatmaps using Ultralytics YOLOv8 alongside Object Counting 🔥🔥🔥

**Comment by yeongnamtan** [comment]
How do we change the colour of the line counter ?

**Reply by RizwanMunawar** [reply]
@yeongnamtan you can use `count_reg_color` to change the Line or Region counter's color. i.e
``` python
from ultralytics import YOLO
from ultralytics.solutions import heatmap
import cv2

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"

# Video writer
video_writer = cv2.VideoWriter("heatmap_output.avi",
                               cv2.VideoWriter_fourcc(*'mp4v'),
                               int(cap.get(5)),
                               (int(cap.get(3)), int(cap.get(4))))

line_points = [(256, 409), (694, 532)]  # line for object counting

# Init heatmap
heatmap_obj = heatmap.Heatmap()
heatmap_obj.set_args(colormap=cv2.COLORMAP_PARULA ,
                     imw=cap.get(4),  # should same as cap height
                     imh=cap.get(3),  # should same as cap width
                     view_img=True,
                     shape="circle",
                     count_reg_color=(0, 255, 0)  # change line color according to your needs
                     count_reg_pts=line_points)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break
    tracks = model.track(im0, persist=True, show=False)

    im0 = heatmap_obj.generate_heatmap(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

Thanks
Ultralytics Team!

**Reply by yeongnamtan** [reply]
thank you @RizwanMunawar

**Comment by yasirahmeds** [comment]
how can I use heatmaps in image?

**Reply by RizwanMunawar** [reply]
@yasirahmeds you can utilize the mentioned code below.

```python
from ultralytics import YOLO
from ultralytics.solutions import heatmap
import cv2

model = YOLO("yolov8s.pt")   # YOLOv8 custom/pretrained model

im0 = cv2.imread("path/to/image.png")  # path to image file

# Heatmap Init
heatmap_obj = heatmap.Heatmap()
heatmap_obj.set_args(colormap=cv2.COLORMAP_PARULA ,
                             imw=im0.shape[0],  # should same as im0 width
                             imh=im0.shape[1],  # should same as im0 height
                             view_img=True,
                             shape="circle")


results = model.track(im0, persist=True)
im0 = heatmap_obj.generate_heatmap(im0, tracks=results)
cv2.imwrite("ultralytics_output.png", im0)
```

Thanks
Ultralytics Team!

**Comment by yeongnamtan** [comment]
I get below error when running the code.   Could you advise how to fix it ?

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-01a24ecc6de6> in <module>
     44     tracks = model.track(im0, classes=[0], persist=True, show=False, tracker="bytetrack.yaml", conf=0.3, stream=False)  # class 0 for fish
     45 
---> 46     im0 = heatmap_obj.generate_heatmap(im0, tracks)
     47     video_writer.write(im0)
     48 

/opt/conda/lib/python3.8/site-packages/ultralytics/solutions/heatmap.py in generate_heatmap(self, im0, tracks)
    239             incount_label = 'InCount : ' + f'{self.in_counts}'
    240             outcount_label = 'OutCount : ' + f'{self.out_counts}'
--> 241             self.annotator.count_labels(in_count=incount_label,
    242                                         out_count=outcount_label,
    243                                         count_txt_size=self.count_txt_thickness,

TypeError: count_labels() got an unexpected keyword argument 'in_count'

**Reply by RizwanMunawar** [reply]
@yeongnamtan thanks for reporting the issue. please install `ultralytics==8.0.236` with the mentioned command below and this issue will be fixed.

```pip install ultralytics==8.0.236`

Thanks
Ultralytics Team!

**Reply by yeongnamtan** [reply]
thank you @RizwanMunawar

**Comment by DaixinX** [comment]
why?
python heatmap.py 
Traceback (most recent call last):
  File "/mnt/yolov8.2/heatmap.py", line 2, in <module>
    from ultralytics.solutions import heatmap
ModuleNotFoundError: No module named 'ultralytics.solutions'
I did pip insatll ultralytics

**Reply by RizwanMunawar** [reply]
@DaixinX, ensure that you have installed the latest version of Ultralytics by using the following command:

```pip install -U ultralytics```

I believe this will address your issue.

Thank you,
The Ultralytics Team!
